R/methods.R

```diff
@@ -346,19 +357,32 @@ cooks.distance.nonprobsvy <- function(model,
 hatvalues.nonprobsvy <- function(model,
                                  ...) { # TODO reduce execution time and glm.fit object and customise to variable selection
   if (any(c("nonprobsvy_dr", "nonprobsvy_ipw") %in% class(model))) {
+    X <- model$X
     propensity_scores <- model$prob
     W <- Matrix::Diagonal(x = propensity_scores * (1 - propensity_scores))
-    XWX_inv <-  solve(t(model$X) %*% W %*% model$X)
-    hat_values_sel <- vector(mode = "numeric", length = length(propensity_scores))
-
-    for (i in 1:length(hat_values_sel)) {
-      hat_values_sel[i] <- W[i,i] * model$X[i,] %*% XWX_inv %*% model$X[i,]
-    }
-    #hats <- Matrix::Diagonal(x = W %*% object$X %*% XWX_inv %*% t(object$X))
-    #names(hat_values) <- row.names(model$parameters)
+    H <- as.matrix(sqrt(W) %*% X %*% solve(t(X) %*% W %*% X) %*% t(X) %*% sqrt(W))
```

Comment:
Also can you not just call `hatvalues` method on original engine like `hatvalues.glm` on `glm`'s?